### PR TITLE
Combine same field values with 'In' instead of 'Or'

### PR DESCRIPTION
**Pull request by KevinCranmer**

### CODE REVIEW

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 1]
```diff
@@ -102,15 +97,54 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+          } else {
+            specification = specification.or(orMe);
+          }
         }
       }
       return this;
     }
+
+    MatchesAnySpecifications<E> add(MatchesAnySpecifications<E> mas) {
+      mas.silSpecifications.values().forEach(this::add);
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:138]: include a null check for `mas`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 2]
```diff
@@ -102,15 +97,54 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+      add(mas.specification);
+      return this;
+    }
+
+    Specification<E> toSpecification() {
+      if (!CollectionUtils.isEmpty(silSpecifications)) {
+        silSpecifications.values().forEach(specification::or);
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:145]: `silSpecifications` is guaranteed to be non-null, so you don't need the emptiness check; `specification` may be null though

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 3]
```diff
@@ -102,15 +97,54 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
     @Getter Specification<E> specification;
 
+    Map<String, SelectInListSpecification<E>> silSpecifications = new HashMap<>();
+
     MatchesAnySpecifications<E> add(Specification<E> orMe) {
       if (orMe != null) {
+        if (orMe instanceof SelectInListSpecification<E> addIn) {
+          if (silSpecifications.get(addIn.fieldName()) != null) {
+            addIn =
+                new SelectInListSpecification<>(
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:107]: Can we move this into an `add(silSpec)` method on `SpecificationInList`?

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 4]
```diff
@@ -102,15 +97,54 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
   private static class MatchesAnySpecifications<E> {
     @Getter Specification<E> specification;
 
+    Map<String, SelectInListSpecification<E>> silSpecifications = new HashMap<>();
+
     MatchesAnySpecifications<E> add(Specification<E> orMe) {
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:102]: `orMe` is no longer an appropriate parameter name

File: src/main/java/gov/va/api/lighthouse/vulcan/SelectInListSpecification.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import org.springframework.data.jpa.domain.Specification;
+
+@Getter
+@Setter
+@AllArgsConstructor
+public class SelectInListSpecification<T> implements Specification<T> {
+  private final @NonNull String fieldName;
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/SelectInListSpecification.java:18]: Put the annotation out front

File: src/main/java/gov/va/api/lighthouse/vulcan/SelectInListSpecification.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import org.springframework.data.jpa.domain.Specification;
+
+@Getter
+@Setter
+@AllArgsConstructor
+public class SelectInListSpecification<T> implements Specification<T> {
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/SelectInListSpecification.java:17]: This class can be `@Value`

Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/SelectInListSpecification.java:17]: This can be another inner class in `Specifications`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 7]
```diff
@@ -102,15 +97,54 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
       if (orMe != null) {
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:103]: We can simplify this flow more:
```
if (other == null) {
  return this;
}
if (other instanceof SelectInListSpecification<E> silSpec) {
  // add silSpec to map
  return this;
}
specification = specification != null ? specification.or(other) : other;
return this;
```

**Review by jblefkove-va — CHANGES_REQUESTED** [review]

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java (hunk 3 above)
Comment (KevinCranmer) [reply src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:114]: I added `append()` to `MatchesAnySpecifications` because I didn't want to add another `add()` method since others with other purposes already exist. And I'm not sure what class was meant by `SpecificationInList`, I don't think we want it in `SetInListSpecification` because we're adding to a variable in `MatchesAnySpecifications`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java (hunk 7 above)
Comment (KevinCranmer) [reply src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java]: I had to add some extra stuff to use `SelectInListSpecification` and idk if the `?` syntax is worth it now because it's a little messy

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java (hunk 2 above)
Comment (KevinCranmer) [reply src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:157]: Removing emptiness check causes a test to fail:
```
java.lang.NullPointerException
        at gov.va.api.lighthouse.vulcan.SpecificationsTest.collectAny(SpecificationsTest.java:50)
```

Comment (jblefkove-va) [reply src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:157]: That NPE is from `specification` being null though, not `silSpecifications`. Also, looking at this again, I realize that this method is modifying `specification`, which is also not ideal. How about:

```
var result = specification;
for (var s : silSpecifications.values()) {
  result = s.or(result);
}
return result;
```

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 8]
```diff
@@ -102,15 +104,82 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+        specification = other;
+        return this;
       }
+      specification =
+          specification instanceof SelectInListSpecification<E> spec
+                  && other instanceof SelectInListSpecification<E> addIn
+                  && addIn.fieldName().equals(spec.fieldName())
+              ? new SelectInListSpecification<>(
+                  spec.fieldName(),
+                  Stream.concat(addIn.values().stream(), spec.values().stream())
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:126]: You don't need this check, it is a duplicate of the `append()` method. In essence, if the parameter spec is a `SelectInListSpecification()`, you execute `append()` above to update the map and then return; otherwise, execute the existing OR logic

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java (hunk 3 above)
Comment (jblefkove-va) [reply src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java]: To clarify, this is what I envisioned:

```
var mapped = silSpecifications.get(addIn.fieldName());
silSpecifications.put(addIn.fieldName(), addIn.add(mapped));
```

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 9]
```diff
@@ -102,15 +103,68 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
   private static class MatchesAnySpecifications<E> {
     @Getter Specification<E> specification;
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:104]: Move `@Getter` to the class-level so it also applies to `silSpecifications`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 10]
```diff
@@ -102,15 +103,68 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+        silSpecifications.put(addIn.fieldName(), addIn.add(mapped));
+        return this;
+      }
+      specification = specification != null ? specification.or(other) : other;
+      return this;
+    }
+
+    MatchesAnySpecifications<E> add(MatchesAnySpecifications<E> mas) {
+      if (mas != null) {
+        mas.silSpecifications.values().forEach(this::add);
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:123]: Use getters for accessing the fields on `mas`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 11]
```diff
@@ -102,15 +103,68 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+  @Getter
+  @Setter
+  @AllArgsConstructor
+  @Value
+  private static class SelectInListSpecification<T> implements Specification<T> {
+    @NonNull String fieldName;
+
+    @NonNull Collection<?> values;
+
+    public SelectInListSpecification<T> add(SelectInListSpecification<T> silSpec) {
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:147]: `add(other)` for the parameter name to match `MatchesAnySpecification`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 12]
```diff
@@ -102,15 +103,68 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+      if (silSpec != null) {
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:148]: This method is cleaner with the conditional swapped, i.e. `if parameter is null, bail immediately, otherwise continue building new SelectInListSpecification`

File: src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java [hunk 13]
```diff
@@ -102,15 +103,68 @@ MatchesAllSpecifications<E> add(Specification<E> andMe) {
...
+    }
+
+    @Override
+    public Predicate toPredicate(
+        Root<T> root, CriteriaQuery<?> query, CriteriaBuilder criteriaBuilder) {
+      if (values().size() == 1) {
+        return criteriaBuilder.equal(
+            root.get(fieldName), values.stream().findFirst().orElseThrow());
+      } else {
+        CriteriaBuilder.In<Object> in = criteriaBuilder.in(root.get(fieldName));
```
Comment (jblefkove-va) [inline src/main/java/gov/va/api/lighthouse/vulcan/Specifications.java:164]: var

File: src/test/java/gov/va/api/lighthouse/vulcan/SpecificationsTest.java [hunk 14]
```diff
@@ -72,4 +67,9 @@ void collectAny() {
     verify(all).or(specs.get(1));
     verify(all).or(specs.get(0));
   }
+
+  @Test
+  void selectInList() {
+    assertThat(Specifications.selectInList("category", List.of())).isNull();
```
Comment (jblefkove-va) [inline src/test/java/gov/va/api/lighthouse/vulcan/SpecificationsTest.java:73]: Do we also want a test for the non-null case?
